actionpack/lib/action_dispatch/http/query_parser.rb

```diff
@@ -0,0 +1,215 @@
+    end
+
+    def new_space_limit(key_space_limit)
+      self.class.new @params_class, key_space_limit, param_depth_limit
+    end
+
+    def new_depth_limit(param_depth_limit)
+      self.class.new @params_class, key_space_limit, param_depth_limit
+    end
+
+    private
+      def params_hash_type?(obj)
+        obj.kind_of?(@params_class)
+      end
+
+      def params_hash_has_key?(hash, key)
+        return false if /\[\]/.match?(key)
+
+        key.split(/[\[\]]+/).inject(hash) do |h, part|
+          next h if part == ""
+          return false unless params_hash_type?(h) && h.key?(part)
+          h[part]
+        end
+
+        true
+      end
+
+      def unescape(s)
+        ::Rack::Utils.unescape(s)
+      end
+
+      class Params
+        def initialize(limit)
+          @limit  = limit
+          @size   = 0
+          @params = {}
+        end
+
+        def [](key)
+          @params[key]
+        end
+
+        def []=(key, value)
+          @size += key.size if key && !@params.key?(key)
+          raise ParamsTooDeepError, "exceeded available parameter key space" if @size > @limit
+          @params[key] = value
+        end
+
+        def key?(key)
+          @params.key?(key)
+        end
+
+        # Recursively unwraps nested `Params` objects and constructs an object
+        # of the same shape, but using the objects' internal representations
+        # (Ruby hashes) in place of the objects. The result is a hash consisting
+        # purely of Ruby primitives.
+        #
+        #   Mutation warning!
+        #
+        #   1. This method mutates the internal representation of the `Params`
+        #      objects in order to save object allocations.
+        #
+        #   2. The value you get back is a reference to the internal hash
+        #      representation, not a copy.
+        #
+        #   3. Because the `Params` object's internal representation is mutable
+        #      through the `#[]=` method, it is not thread safe. The result of
+        #      getting the hash representation while another thread is adding a
+        #      key to it is non-deterministic.
+        #
+        def to_h
+          @params.each do |key, value|
+            case value
+            when self
+              # Handle circular references gracefully.
+              @params[key] = @params
+            when Params
+              @params[key] = value.to_h
+            when Array
+              value.map! { |v| v.kind_of?(Params) ? v.to_h : v }
+            else
+              # Ignore anything that is not a `Params` object or
+              # a collection that can contain one.
+            end
+          end
+          @params
+        end
+        alias_method :to_params_hash, :to_h
+      end
+  end
+end
+
+# Replace the Rack query parser with the non-standard Rack 2.2 implementation.
+# See <https://github.com/rails/rails/pull/47080> for more details.
+Rack::Utils.default_query_parser = ActionDispatch::QueryParser.make_default
```

Comment:
Copying the class in here looks good, but this feels risky.

Instead of changing Rack's default, can we arrange to use our parser when we're doing the parsing (and/or most importantly, in the production of our `params` value), but leave Rack::Request to do whatever Rack wants if e.g. a third-party middleware is calling it directly?

Reply:
No that’s not possible because the input body can be consumed and cached by `Rack::Request`. The parameters only get parsed once and it will become non-determine whether rack or rails parsers get to it first in your proposed change.

Reply:
Ah, and we can't separately cache our parsed interpretation, because the input isn't rewindable.

By even calling it "default", then, it seems that Rack's API is implying a level of flexibility it can not support (any more?).

An effective global monkey-patch of Rack behaviour as a result of loading our library is not an acceptable outcome... at best that will create a per-project non-determinism that will confound any Rack-downstream library, making them behave differently when loaded in a Rails app.

I think my current best suggestion is that we install a middleware to make the input rewindable [only] if it has a relevant content-type -- after which we should be able to do our own parameter parsing independently of rack, caching separately.

Reply:
`default_query_parser` is designed for exactly this scenario and there is no problem changing it, but as I said we should probably make this user configurable.

Rack has lots of globally controllable behaviour and that's just the nature of it.

> I think my current best suggestion is that we install a middleware to make the input rewindable [only] if it has a relevant content-type -- after which we should be able to do our own parameter parsing independently of rack, caching separately.

I don't think this is a practical solution, and carries a high level of complexity and technical debt.

Reply:
I'd really like to find a solution to this.  It'll be very confusing if parsing is different depending on what files are loaded.  I can imagine a scenario where someone is debugging QP parsing in a Rails app (I have had to do this) and they do it via the Rack APIs, but it's not reflective of reality since this file will be loaded in the app but not necessarily your IRB session.

> No that’s not possible because the input body can be consumed and cached by Rack::Request.

If this is the case, then no `Rack::Request` object should be allowed to be allocated before the request object in Rails, no?  Otherwise our request object in Rails would always see empty QPs on POST.

Reply:
> If this is the case, then no Rack::Request object should be allowed to be allocated before the request object in Rails, no?

How do you propose to prevent this?

Reply:
> How do you propose to prevent this?

I don't think we can.  But if users add a middleware that uses `Rack::Request`, surely they'll notice their QPs are gone and not use that middleware?

I guess this raises a larger issues:  We basically require anyone using a `Rack::Request` object to cache any calculations as subsequent allocations _won't work_ since the body isn't rewindable.

Since allocating a Rack::Request isn't idempotent (it reads the body, mutating the input stream), it seems like we should probably make the request object a singleton, and the type configurable (Rails wants to have it's own request object).

Side note: I am starting to feel great regret for changing the way QPs are parsed in Rack!